### PR TITLE
Bring Stats API up to speed: Add `conversion_rate` to Aggregate and Breakdown

### PR DESCRIPTION
**Pull request by RobertJoonas**
### Changes

**For the reviewers:** Reviewing commit-by-commit highly recommended

The main objective of this PR is to add the `conversion_rate` metric to the aggregate and breakdown endpoints of Stats API.

In addition to that, since I was already working on the conversion report in different breakdown reports, I made the following UI fixes as well:

* Display conversion rate in Regions and Cities (ListReport view) - The conversion rate was only missing from these two breakdown reports - all others, such as Browsers, Browser Versions, etc had it already.

Before (prod now):
image

After (from staging):
image

* Display total visitors, conversions (visitors), and CR in the "Details" modals of Countries, Regions, and Cities. These are the metrics in all other detailed breakdown reports on the dashboard when a goal filter is applied.

Before (prod):
image

After (staging):
image

* Move the percentage into a separate column in the Countries details table (goal filter **not** applied):

Before (prod):
image

After (staging):
image

### Tests
- [x] Automated tests have been added

### Changelog
- [ ] Entry has been added to changelog

### Documentation
- [x] PR ready https://github.com/plausible/docs/pull/461

### Dark mode
- [x] This PR does not change the UI

### CODE REVIEW

**Comment by metmarkosaric** [comment]
Very nice @RobertJoonas! How about adding percentage columns to regions and cities too while you're at it to keep it all consistent? 😁

**Comment by RobertJoonas** [comment]
> How about adding percentage columns to regions and cities too while you're at it to keep it all consistent?

@metmarkosaric I was thinking about that actually, but there's a bug with percentage metric that was recently attempted to be solved but I see it's still happening in Sentry. So I'd like to avoid touching that part at this point and move forward with this PR.

**Comment by metmarkosaric** [comment]
ahh ok sounds good! this is a great improvement already! 👏

File: lib/plausible_web/controllers/api/stats_controller.ex [hunk 1]
```diff
@@ -1075,16 +1075,7 @@ defmodule PlausibleWeb.Api.StatsController do
...
-      on_full_build do
-        if Enum.any?(site.goals, &Plausible.Goal.Revenue.revenue?/1) do
-          [:visitors, :events] ++ @revenue_metrics
-        else
-          [:visitors, :events]
-        end
-      else
-        [:visitors, :events]
-      end
+    metrics = [:visitors, :events] ++ @revenue_metrics
```
[the remark this answers is not in the document]
Comment (RobertJoonas) [reply lib/plausible_web/controllers/api/stats_controller.ex]: I ended up keeping my version of the changes in `stats_controller.ex` in the [merge commit](https://github.com/plausible/analytics/pull/3739/commits/e550481ea4ce2a39cd211dc14c5bd9f7399351a5), but I applied the other logic (restricting revenue goal access without plan access) in a separate commit into the lower level breakdown API: https://github.com/plausible/analytics/pull/3739/commits/9f3d9997f67726027406809dcf5889ae8cdb41e3

File: lib/plausible_web/controllers/api/external_stats_controller.ex [hunk 2]
```diff
@@ -113,6 +117,9 @@ defmodule PlausibleWeb.Api.ExternalStatsController do
         {:error, reason}
 
       metrics ->
+        # Only call `String.to_existing_atom/1` to convert atoms
+        # existing in the current module (a generally good practice).
+        _existing = @stats_api_metrics
```
[the remark this answers is not in the document]
Comment (zoldar) [reply lib/plausible_web/controllers/api/external_stats_controller.ex]: https://github.com/plausible/analytics/pull/3739/commits/07406d00eb71f8f016f4cb0b186a04a62c7e3118

File: lib/plausible_web/controllers/api/stats_controller.ex [hunk 3]
```diff
@@ -340,61 +340,23 @@ defmodule PlausibleWeb.Api.StatsController do
...
-        nil
-      end
+    results = Stats.aggregate(site, query, metrics)
+    comparison = if comparison_query, do: Stats.aggregate(site, comparison_query, metrics)
 
     [
-      top_stats_entry(results, comparison, "Unique visitors", :unique_visitors),
-      top_stats_entry(results, comparison, "Unique conversions", :converted_visitors),
-      top_stats_entry(results, comparison, "Total conversions", :completions),
+      top_stats_entry(results, comparison, "Unique visitors", :visitors_without_event_filters),
```
Comment (zoldar) [inline lib/plausible_web/controllers/api/stats_controller.ex:350]: Perhaps rename this to `:total_visitors`?

File: lib/plausible/stats/aggregate.ex [hunk 4]
```diff
@@ -40,12 +45,38 @@ defmodule Plausible.Stats.Aggregate do
 
     Plausible.ClickhouseRepo.parallel_tasks([session_task, event_task, time_on_page_task])
     |> Enum.reduce(%{}, fn aggregate, task_result -> Map.merge(aggregate, task_result) end)
+    |> maybe_put_cr(site, query, metrics)
+    |> maybe_remove_visitors_metric(metrics)
```
Comment (ukutaht) [inline lib/plausible/stats/aggregate.ex:49]: Suggestion: refactor this function to `keep_requested_metrics(metrics)` or something like this which would remove any metric from results that wasn't requested. That could also replace the `remove_internal_visits` function.

File: lib/plausible/stats/aggregate.ex [hunk 5]
```diff
@@ -26,8 +26,13 @@ defmodule Plausible.Stats.Aggregate do
 
     Query.trace(query, metrics)
 
-    event_metrics = Enum.filter(metrics, &(&1 in @event_metrics))
+    event_metrics =
+      metrics
+      |> maybe_add_visitors_metric()
```
[the remark this answers is not in the document]
Comment (RobertJoonas) [reply lib/plausible/stats/aggregate.ex]: ✅ https://github.com/plausible/analytics/pull/3739/commits/de0517feb66b49d57a83e41f7757c8d4a991b5f7

File: lib/plausible/stats/breakdown.ex [hunk 6]
```diff
@@ -29,15 +29,25 @@ defmodule Plausible.Stats.Breakdown do
 
     if !Keyword.get(opts, :skip_tracing), do: trace(query, property, metrics)
 
+    {revenue_goals, metrics} =
+      on_full_build do
+        revenue_goals = Enum.filter(event_goals, &Plausible.Goal.Revenue.revenue?/1)
+        metrics = if Enum.any?(revenue_goals), do: metrics, else: metrics -- @revenue_metrics
```
[the remark this answers is not in the document]
Comment (RobertJoonas) [reply lib/plausible/stats/breakdown.ex]: Makes sense! Done in https://github.com/plausible/analytics/pull/3739/commits/39d35f8337de534421a11deb561587f9431eafb3

File: lib/plausible/stats/breakdown.ex [hunk 7]
```diff
@@ -29,15 +29,25 @@ defmodule Plausible.Stats.Breakdown do
...
+
+        {revenue_goals, metrics}
+      else
+        {nil, metrics}
+      end
+
+    metrics_to_select =
+      metrics
+      |> maybe_add_visitors_metric()
+      |> Kernel.--(@special_metrics)
```
[the remark this answers is not in the document]
Comment (RobertJoonas) [reply lib/plausible/stats/breakdown.ex]: Agreed! https://github.com/plausible/analytics/pull/3739/commits/b1c108d759ecbe1856ef14607d596839495b6f9c

File: lib/plausible/stats/breakdown.ex [hunk 8]
```diff
@@ -16,7 +16,7 @@ defmodule Plausible.Stats.Breakdown do
 
   @event_metrics [:visitors, :pageviews, :events] ++ @revenue_metrics
 
-  @event_props Plausible.Stats.Props.event_props()
+  @special_metrics [:conversion_rate, :total_visitors]
```
[the remark this answers is not in the document]
Comment (RobertJoonas) [reply lib/plausible/stats/breakdown.ex]: Right... `computed_metrics` sounds much better! Done in https://github.com/plausible/analytics/pull/3739/commits/2f884e7c9813d9b5a6d1ae4526aacc72b2b3d3a6

File: lib/plausible_web/controllers/api/stats_controller.ex (hunk 3 above)
Comment (RobertJoonas) [reply lib/plausible_web/controllers/api/stats_controller.ex]: Totally. https://github.com/plausible/analytics/pull/3739/commits/cb0b7744db4602a2f359c03ea35dfa88f599542a

File: lib/plausible/stats/aggregate.ex (hunk 4 above)
Comment (RobertJoonas) [reply lib/plausible/stats/aggregate.ex]: Yeah, I like that. I tried to just `Map.take` the requested metrics at first, but the problem is that the breakdown property is included in the same map, along with the metrics. E.g:

```elixir
%{browser: "Chrome", visitors: 123, pageviews: 234}
```

What I've done is define a list of metrics to "maybe drop" from the results map. let me know what you think https://github.com/plausible/analytics/pull/3739/commits/2d2ff40294cad951ae30e042101c6ad69ef95c6b

**Pull request merged**